### PR TITLE
settings: set `access-control-allow-origin: *` header (bug 2030883)

### PR DESCRIPTION
**Pull request by shtrom**
TreeHerder needs to be allowed to make requests to the `/landing_jobs/<job_id>` JSON endpoint to determine the landed Hg commit_id for a given job.

 * deps: add django-cors-headers
 * settings: set `access-control-allow-origin: *` header
 * test: add middlewares test for ACAO header

### CODE REVIEW

File: src/lando/main/tests/test_models.py [hunk 1]
```diff
@@ -87,14 +87,18 @@ def test__models__Repo__scm_not_calculated_when_preset(
         ("/invalid_path", ValidationError),
     ],
 )
-def test__models__Repo__system_path_validator(path, expected_exception):
+def test__models__Repo__system_path_validator(
+    request: pytest.FixtureRequest, path: str, expected_exception: Exception | None
```
Comment (shtrom) [inline src/lando/main/tests/test_models.py]: Without this change, this test that passed locally failed in GhA. It is absolutely unclear why it makes any difference.

I suspect the change just allowed some cache-busting to happen.

File: src/lando/tests/test_middlewares.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+    resp = client.get(path, headers=headers)
+
+    if expected_present:
+        assert (
+            "access-control-allow-origin" in resp.headers
+        ), f"Missing ACAO header for request from '{origin}' to '{path}'"
+        assert (
+            resp.headers["access-control-allow-origin"] == "*"
+        ), f"Unexpected ACAO header value {resp.headers["access-control-allow-origin"]} for requesst from '{origin}' to '{path}'"
```
Comment (cgsheeh) [inline src/lando/tests/test_middlewares.py]:
```suggestion
        ), f"Unexpected ACAO header value {resp.headers["access-control-allow-origin"]} for request from '{origin}' to '{path}'"
```

File: src/lando/tests/test_middlewares.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    else:
+        assert (
+            "access-control-allow-origin" not in resp.headers
+        ), f"Unexpected ACAO header present for requesst from '{origin}' to '{path}'"
```
Comment (cgsheeh) [inline src/lando/tests/test_middlewares.py]:
```suggestion
        ), f"Unexpected ACAO header present for request from '{origin}' to '{path}'"
```

File: src/lando/tests/test_middlewares.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        ("treeherder", "/api/1", False),
+    ),
+)
+@pytest.mark.django_db
+def test_cors_acao_header(
+    client: Client, origin: str, path: str, expected_present: bool
+):
+
+    headers = {}
+    if origin:
```
Comment (cgsheeh) [inline src/lando/tests/test_middlewares.py]: Perhaps the "no origin" case should be `None`, in case we need to test the empty string? So this would be `if origin is not None:`?

File: src/lando/main/tests/test_models.py (hunk 1 above)
Comment (cgsheeh) [reply src/lando/main/tests/test_models.py]: Perhaps related to `pytest-xdist`? AIUI adding fixtures can change how tests are distributed. We have a few race conditions in tests that can cause intermittents locally, perhaps you triggered one and adding this line changes the test run order such that the intermittent is no longer triggered. What was the error you saw?

I should file bugs for the intermittent failures, will do when they pop up again. 👍

**Review by cgsheeh — APPROVED** [review]

File: src/lando/main/tests/test_models.py (hunk 1 above)
Comment (shtrom) [reply src/lando/main/tests/test_models.py]: I've noticed one fairly regular race in one HgMO test, definitely due to xdist, but not frequent enough to really be a bother.

<details>

<summary>
Oh! It just happened. Noting it down here until the bug is filed, but completely unrelated to this
</summary>

```
_ test_merge_conflict[hg-hunks FAILED-# HG changeset patch\n# User Test User <test@example.com>\n# Date 0 0\n#      Thu Jan 01 00:00:00 1970 +0000\n# Diff Start Line 7\nno bug: add another file for formatting 2\n\ndiff --git a/test.txt b/test.txt\n--- a/test.txt\n+++ b/test.txt\n@@ -2,3 +2,4 @@ TEST\n\n \n adding another line\n+add one more line\n] _
[gw4] linux -- Python 3.12.13 /usr/local/bin/python3.12

repo_mc = <function repo_mc.<locals>.factory at 0x7fd4ce1476a0>
treestatusdouble = <lando.api.tests.mocks.TreeStatusDouble object at 0x7fd4ce0fdc70>
mock_phab_trigger_repo_update_apply_async = <MagicMock id='140551966945696'>
create_patch_revision = <function create_patch_revision.<locals>._create_patch_revision at 0x7fd4ce272de0>
make_landing_job = <function make_landing_job.<locals>.landing_job_factory at 0x7fd4ce271300>
caplog = <_pytest.logging.LogCaptureFixture object at 0x7fd4ce03eab0>
get_landing_worker = <function get_landing_worker.<locals>._get_landing_worker at 0x7fd4ce16d4e0>
repo_type = SCMType.HG, expected_error_log = 'hunks FAILED'
patch = '# HG changeset patch\n# User Test User <test@example.com>\n# Date 0 0\n#      Thu Jan 01 00:00:00 1970 +0000\n# Diff ....txt b/test.txt\n--- a/test.txt\n+++ b/test.txt\n@@ -2,3 +2,4 @@ TEST\n\n \n adding another line\n+add one more line\n'

    @pytest.mark.parametrize(
        "repo_type, expected_error_log, patch",
        # Can't use itertools.product without similar overhead as this,
        # as we have more than one element in the first set.
        (
            scm + (patch,)
            for scm in (
                (SCMType.GIT, "Rejected hunk"),
                (SCMType.HG, "hunks FAILED"),
            )
            for patch in (
                PATCH_FORMATTED_2,
                PATCH_CHANGE_MISSING_CONTENT,
            )
        ),
    )
    @pytest.mark.django_db
    def test_merge_conflict(
        repo_mc: Callable,
        treestatusdouble: TreeStatusDouble,
        mock_phab_trigger_repo_update_apply_async: mock.Mock,
        create_patch_revision: Callable,
        make_landing_job: Callable,
        caplog: pytest.LogCaptureFixture,
        get_landing_worker: Callable,
        repo_type: str,
        expected_error_log: str,
        patch: str,
    ):
        repo = repo_mc(repo_type)
        treestatusdouble.open_tree(repo.name)

        job_params = {
            "id": 1234,
            "status": JobStatus.IN_PROGRESS,
            "requester_email": "test@example.com",
            "target_repo": repo,
            "attempts": 1,
        }
        job = make_landing_job(
            revisions=[
                create_patch_revision(1, patch=patch),
            ],
            **job_params,
        )

        worker = get_landing_worker(repo_type)
        assert worker.run_job(job)
        assert job.status == JobStatus.FAILED

        assert expected_error_log in caplog.text

        assert job.error_breakdown, "No error breakdown added to job"
        assert job.error_breakdown.get(
            "rejects_paths"
        ), "Empty or missing reject information in error breakdown"
        failed_paths = [p["path"] for p in job.error_breakdown["failed_paths"]]
        assert set(failed_paths) == set(
            job.error_breakdown["rejects_paths"].keys()
        ), "Mismatch between failed_paths and rejects_paths"
        for fp in failed_paths:
            assert job.error_breakdown["rejects_paths"][fp].get(
                "path"
            ), f"Empty or missing reject path for failed path {fp}"
>           assert job.error_breakdown["rejects_paths"][fp].get(
                "content"
            ), f"Empty or missing reject content for failed path {fp}"
E           AssertionError: Empty or missing reject content for failed path test.txt
E           assert None
E            +  where None = <built-in method get of dict object at 0x7fd4ceabd200>('content')
E            +    where <built-in method get of dict object at 0x7fd4ceabd200> = {'path': 'test.txt.rej'}.get

src/lando/api/tests/test_landings.py:809: AssertionError
----------------------------------------------- Captured stdout setup -----------------------------------------------
Initialized empty Git repository in /tmp/pytest-of-root/pytest-0/popen-gw4/test_merge_conflict_hg_hunks_F0/git_repo/.git/
Applying: initial commit
Applying: add another file
[dev d593084] dev
 Date: Thu Jan 1 00:00:00 1970 +0000
requesting all changes
adding changesets
adding manifests
adding file changes
added 2 changesets with 2 changes to 2 files
new changesets 7048446d5acc:0da79df0ffff (2 drafts)
updating to branch default
2 files updated, 0 files merged, 0 files removed, 0 files unresolved
127.0.0.1 - - [14/Apr/2026 02:22:48] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [14/Apr/2026 02:22:48] "GET /?cmd=capabilities HTTP/1.1" 200 -
127.0.0.1 - - [14/Apr/2026 02:22:48] "GET /?cmd=batch HTTP/1.1" 200 - x-hgarg-1:cmds=heads+%3Bknown+nodes%3D x-hgproto-1:0.1 0.2 comp=zstd,zlib,none,bzip2 partial-pull
requesting all changes
127.0.0.1 - - [14/Apr/2026 02:22:48] "GET /?cmd=getbundle HTTP/1.1" 200 - x-hgarg-1:bookmarks=1&bundlecaps=HG20%2Cbundle2%3DHG20%250Abookmarks%250Achangegroup%253D01%252C02%250Acheckheads%253Drelated%250Adigests%253Dmd5%252Csha1%252Csha512%250Aerror%253Dabort%252Cunsupportedcontent%252Cpushraced%252Cpushkey%250Ahgtagsfnodes%250Alistkeys%250Aphases%253Dheads%250Apushkey%250Aremote-changegroup%253Dhttp%252Chttps%250Astream%253Dv2&cg=1&common=0000000000000000000000000000000000000000&heads=0da79df0ffff88e0ad6fa3e27508bcf5b2f2cec4&listkeys=bookmarks&phases=1 x-hgproto-1:0.1 0.2 comp=zstd,zlib,none,bzip2 partial-pull
adding changesets
adding manifests
adding file changes
added 2 changesets with 2 changes to 2 files
new changesets 7048446d5acc:0da79df0ffff
updating to branch default
2 files updated, 0 files merged, 0 files removed, 0 files unresolved
----------------------------------------------- Captured stderr setup -----------------------------------------------
hint: Using 'master' as the name for the initial branch. This default branch name
hint: is subject to change. To configure the initial branch name to use in all
hint: of your new repositories, which will suppress this warning, call:
hint:
hint: 	git config --global init.defaultBranch <name>
hint:
hint: Names commonly chosen instead of 'master' are 'main', 'trunk' and
hint: 'development'. The just-created branch can be renamed via this command:
hint:
hint: 	git branch -m <name>
applying to an empty history
Switched to a new branch 'dev'
Switched to branch 'main'
----------------------------------------------- Captured stdout call ------------------------------------------------
127.0.0.1 - - [14/Apr/2026 02:22:50] "GET /?cmd=capabilities HTTP/1.1" 200 -
127.0.0.1 - - [14/Apr/2026 02:22:50] "GET /?cmd=lookup HTTP/1.1" 200 - x-hgarg-1:key=default x-hgproto-1:0.1 0.2 comp=zstd,zlib,none,bzip2 partial-pull
127.0.0.1 - - [14/Apr/2026 02:22:50] "GET /?cmd=capabilities HTTP/1.1" 200 -
127.0.0.1 - - [14/Apr/2026 02:22:50] "GET /?cmd=batch HTTP/1.1" 200 - x-hgarg-1:cmds=heads+%3Bknown+nodes%3D0da79df0ffff88e0ad6fa3e27508bcf5b2f2cec4 x-hgproto-1:0.1 0.2 comp=zstd,zlib,none,bzip2 partial-pull
127.0.0.1 - - [14/Apr/2026 02:22:50] "GET /?cmd=getbundle HTTP/1.1" 200 - x-hgarg-1:bookmarks=1&bundlecaps=HG20%2Cbundle2%3DHG20%250Abookmarks%250Achangegroup%253D01%252C02%250Acheckheads%253Drelated%250Adigests%253Dmd5%252Csha1%252Csha512%250Aerror%253Dabort%252Cunsupportedcontent%252Cpushraced%252Cpushkey%250Ahgtagsfnodes%250Alistkeys%250Aphases%253Dheads%250Apushkey%250Aremote-changegroup%253Dhttp%252Chttps%250Astream%253Dv2&cg=0&common=0da79df0ffff88e0ad6fa3e27508bcf5b2f2cec4&heads=0da79df0ffff88e0ad6fa3e27508bcf5b2f2cec4&listkeys=bookmarks&phases=1 x-hgproto-1:0.1 0.2 comp=zstd,zlib,none,bzip2 partial-pull
------------------------------------------------- Captured log call -------------------------------------------------
ERROR    lando.main.scm.hg:hg.py:409 [Errno 2] No such file or directory: '/tmp/patch_rejects/tmp/pytest-of-root/pytest-0/popen-gw4/test_merge_conflict_hg_hunks_F0/hg_clone/test.txt.rej'
Traceback (most recent call last):
  File "/code/src/lando/main/scm/hg.py", line 672, in run_hg
    return self._run_hg(args)
           ^^^^^^^^^^^^^^^^^^
  File "/code/src/lando/main/scm/hg.py", line 721, in _run_hg
    raise hglib.error.CommandError(args, ret, out, err)
hglib.error.CommandError: (255, b'applying /tmp/tmpnnkj07d0diff', b'test.txt\nHunk #1 FAILED at 2.\n1 out of 1 hunk FAILED -- saving rejects to file test.txt.rej\nabort: patch command failed: exited with status 1')

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/code/src/lando/main/scm/hg.py", line 335, in _run_hg_patch
    self.run_hg(import_cmd + similarity_args + [patch_or_diff.name])
  File "/code/src/lando/main/scm/hg.py", line 674, in run_hg
    raise HgException.from_hglib_error(exc) from exc
lando.main.scm.hg.HgPatchConflict: hg error in cmd: hg import --no-commit --config ui.patch=patch -s 95 /tmp/tmpnnkj07d0diff: applying /tmp/tmpnnkj07d0diff

test.txt
Hunk #1 FAILED at 2.
1 out of 1 hunk FAILED -- saving rejects to file test.txt.rej
abort: patch command failed: exited with status 1

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/code/src/lando/api/legacy/workers/base.py", line 327, in handle_new_commit_failures
    return create_revision_callable(revision)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/code/src/lando/api/legacy/workers/landing_worker.py", line 291, in apply_patch
    scm.apply_patch(
  File "/code/src/lando/main/scm/hg.py", line 261, in apply_patch
    self._run_hg_patch(import_cmd, f_diff, similarity_args=similarity_args)
  File "/code/src/lando/main/scm/hg.py", line 345, in _run_hg_patch
    raise exc from exc2
  File "/code/src/lando/main/scm/hg.py", line 323, in _run_hg_patch
    self.run_hg(import_cmd + [patch_or_diff.name])
  File "/code/src/lando/main/scm/hg.py", line 674, in run_hg
    raise HgException.from_hglib_error(exc) from exc
lando.main.scm.hg.HgPatchConflict: hg error in cmd: hg import --no-commit /tmp/tmpnnkj07d0diff: applying /tmp/tmpnnkj07d0diff

patching file test.txt
Hunk #1 FAILED at 1
1 out of 1 hunks FAILED -- saving rejects to file test.txt.rej
abort: patch failed to apply

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/code/src/lando/main/scm/hg.py", line 406, in process_merge_conflict
    with open(self._get_rejects_path() / self.path[1:] / path, "r") as f:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/patch_rejects/tmp/pytest-of-root/pytest-0/popen-gw4/test_merge_conflict_hg_hunks_F0/hg_clone/test.txt.rej'
ERROR    lando.api.legacy.workers.base:base.py:351 Problem while applying patch in revision 1:

hg error in cmd: hg import --no-commit /tmp/tmpnnkj07d0diff: applying /tmp/tmpnnkj07d0diff

patching file test.txt
Hunk #1 FAILED at 1
1 out of 1 hunks FAILED -- saving rejects to file test.txt.rej
abort: patch failed to apply
Traceback (most recent call last):
  File "/code/src/lando/main/scm/hg.py", line 672, in run_hg
    return self._run_hg(args)
           ^^^^^^^^^^^^^^^^^^
  File "/code/src/lando/main/scm/hg.py", line 721, in _run_hg
    raise hglib.error.CommandError(args, ret, out, err)
hglib.error.CommandError: (255, b'applying /tmp/tmpnnkj07d0diff', b'test.txt\nHunk #1 FAILED at 2.\n1 out of 1 hunk FAILED -- saving rejects to file test.txt.rej\nabort: patch command failed: exited with status 1')

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/code/src/lando/main/scm/hg.py", line 335, in _run_hg_patch
    self.run_hg(import_cmd + similarity_args + [patch_or_diff.name])
  File "/code/src/lando/main/scm/hg.py", line 674, in run_hg
    raise HgException.from_hglib_error(exc) from exc
lando.main.scm.hg.HgPatchConflict: hg error in cmd: hg import --no-commit --config ui.patch=patch -s 95 /tmp/tmpnnkj07d0diff: applying /tmp/tmpnnkj07d0diff

test.txt
Hunk #1 FAILED at 2.
1 out of 1 hunk FAILED -- saving rejects to file test.txt.rej
abort: patch command failed: exited with status 1

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/code/src/lando/api/legacy/workers/base.py", line 327, in handle_new_commit_failures
    return create_revision_callable(revision)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/code/src/lando/api/legacy/workers/landing_worker.py", line 291, in apply_patch
    scm.apply_patch(
  File "/code/src/lando/main/scm/hg.py", line 261, in apply_patch
    self._run_hg_patch(import_cmd, f_diff, similarity_args=similarity_args)
  File "/code/src/lando/main/scm/hg.py", line 345, in _run_hg_patch
    raise exc from exc2
  File "/code/src/lando/main/scm/hg.py", line 323, in _run_hg_patch
    self.run_hg(import_cmd + [patch_or_diff.name])
  File "/code/src/lando/main/scm/hg.py", line 674, in run_hg
    raise HgException.from_hglib_error(exc) from exc
lando.main.scm.hg.HgPatchConflict: hg error in cmd: hg import --no-commit /tmp/tmpnnkj07d0diff: applying /tmp/tmpnnkj07d0diff

patching file test.txt
Hunk #1 FAILED at 1
1 out of 1 hunks FAILED -- saving rejects to file test.txt.rej
abort: patch failed to apply

[...]

FAILED src/lando/api/tests/test_landings.py::test_merge_conflict[hg-hunks FAILED-# HG changeset patch\n# User Test User <test@example.com>\n# Date 0 0\n#      Thu Jan 01 00:00:00 1970 +0000\n# Diff Start Line 7\nno bug: add another file for formatting 2\n\ndiff --git a/test.txt b/test.txt\n--- a/test.txt\n+++ b/test.txt\n@@ -2,3 +2,4 @@ TEST\n\n \n adding another line\n+add one more line\n] - AssertionError: Empty or missing reject content for failed path test.txt
```
</details>

This one seemed to choke on not using the `setting.REPO_ROOT` value, which the test overwrite to match where the tests create the repo... But not in the right order? But then again, changing type-hints fixed the order, so.... 🤷🏼 ❓
